src/abstract_cost.jl

```diff
@@ -1,76 +1,72 @@
+@inline call(c::PredictionLoss, X, y, α) = value(c, X, y, α)
+@inline transpose(c::PredictionLoss) = deriv_fun(c)
 
-@inline function value_fun(c::Cost)
+@inline function value_fun(c::PredictionLoss)
     @inline _value(args...) = value(c, args...)
     _value
 end
 
-@inline function deriv_fun(c::Cost)
+@inline function deriv_fun(c::PredictionLoss)
     @inline _deriv(args...) = deriv(c, args...)
     _deriv
 end
 
-@inline function deriv2_fun(c::Cost)
+@inline function deriv2_fun(c::PredictionLoss)
     @inline _deriv2(args...) = deriv2(c, args...)
     _deriv2
 end
 
-@inline function value_deriv_fun(c::Cost)
+@inline function value_deriv_fun(c::PredictionLoss)
     @inline _value_deriv(args...) = value_deriv(c, args...)
     _value_deriv
 end
 
-isminimizable(c::Cost) = isconvex(c)
-isdifferentiable(c::Cost) = istwicedifferentiable(c)
-istwicedifferentiable(::Cost) = false
-isdifferentiable(c::Cost, at) = isdifferentiable(c)
-istwicedifferentiable(c::Cost, at) = istwicedifferentiable(c)
-isconvex(::Cost) = false
-isstronglyconvex(::Cost) = false
+isminimizable(c::PredictionLoss) = isconvex(c)
+isdifferentiable(c::PredictionLoss) = istwicedifferentiable(c)
+istwicedifferentiable(::PredictionLoss) = false
+isdifferentiable(c::PredictionLoss, at) = isdifferentiable(c)
+istwicedifferentiable(c::PredictionLoss, at) = istwicedifferentiable(c)
+isconvex(::PredictionLoss) = false
+isstronglyconvex(::PredictionLoss) = false
 
 # ==========================================================================
 
-abstract Loss <: Cost
-
-isnemitski(l::Loss) = islocallylipschitzcont(l)
-islipschitzcont(::Loss) = false
-islocallylipschitzcont(::Loss) = false
-isclipable(::Loss) = false
-islipschitzcont_deriv(::Loss) = false
+isnemitski(l::PredictionLoss) = islocallylipschitzcont(l)
+islipschitzcont(::PredictionLoss) = false
+islocallylipschitzcont(::PredictionLoss) = false
+isclipable(::PredictionLoss) = false
+islipschitzcont_deriv(::PredictionLoss) = false
 
 # ==========================================================================
 
-abstract SupervisedLoss <: Loss
-
-@inline call(l::SupervisedLoss, y, t) = value(l, y, t)
-@inline value(l::SupervisedLoss, X::AbstractArray, y::Number, t::Number) = value(l, y, t)
-@inline deriv(l::SupervisedLoss, X::AbstractArray, y::Number, t::Number) = deriv(l, y, t)
-@inline deriv2(l::SupervisedLoss, X::AbstractArray, y::Number, t::Number) = deriv2(l, y, t)
-@inline value_deriv(l::SupervisedLoss, X::AbstractArray, y::Number, t::Number) = value_deriv(l, y, t)
+@inline call(l::PredictionLoss, y, t) = value(l, y, t)
+@inline value(l::PredictionLoss, X::AbstractArray, y::Number, t::Number) = value(l, y, t)
+@inline deriv(l::PredictionLoss, X::AbstractArray, y::Number, t::Number) = deriv(l, y, t)
+@inline deriv2(l::PredictionLoss, X::AbstractArray, y::Number, t::Number) = deriv2(l, y, t)
+@inline value_deriv(l::PredictionLoss, X::AbstractArray, y::Number, t::Number) = value_deriv(l, y, t)
```

Comment:
If we drop the supervised vs unsupervised then these 4 parameters methods with `X`, `y`, and `t` can be dropped